### PR TITLE
[Merged by Bors] - feat: port Data.ZMod.Defs

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Zmod/Defs.lean [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    rw [← lhs, left_distrib]
+
+-- Porting note: new
+instance (n) [NeZero n] : CommSemiring (Fin n) where
+  __ := inferInstanceAs (MonoidWithZero (Fin n))
+  __ := inferInstanceAs (CommSemigroup (Fin n))
+  __ := inferInstanceAs (AddCommMonoid (Fin n))
+  __ := inferInstanceAs (AddMonoidWithOne (Fin n))
+  left_distrib := Fin.mul_add
+  right_distrib a b c := (by rw [mul_comm, Fin.mul_add, mul_comm c, mul_comm c])
```
Comment (eric-wieser) [inline Mathlib/Data/Zmod/Defs.lean:80]: Why are these new? Don't they exist in mathlib3?

**Review by eric-wieser — COMMENTED** [review]
Did you use the output of mathlib3port or the ad-hoc content here? The intent was for the adhoc code to be discarded entirely.

File: Mathlib/Data/Zmod/Defs.lean [hunk 2]
```diff
@@ -46,37 +46,64 @@ to register the ring structure on `zmod n` as type class instance.
...
-
-private theorem left_distrib_aux (n : ℕ) : ∀ a b c : Fin n, a * (b + c) = a * b + a * c :=
-  fun ⟨a, ha⟩ ⟨b, hb⟩ ⟨c, hc⟩ =>
-  Fin.eq_of_veq
-    (calc
-      a * ((b + c) % n) ≡ a * (b + c) [MOD n] := (Nat.mod_modEq _ _).mul_left _
-      _ ≡ a * b + a * c [MOD n] := by rw [mul_add]
-      _ ≡ a * b % n + a * c % n [MOD n] := (Nat.mod_modEq _ _).symm.add (Nat.mod_modEq _ _).symm
-      )
-#align fin.left_distrib_aux fin.left_distrib_aux
```
Comment (eric-wieser) [inline Mathlib/Data/Zmod/Defs.lean]: I think you should restore these lines, and the old `CommRing` instance

Comment (Ruben-VandeVelde) [reply Mathlib/Data/Zmod/Defs.lean]: There were some errors in the generated code, so it seemed easier to use the working code. I can take a look at fixing them instead, though probably after `CommRing` works

Comment (eric-wieser) [reply Mathlib/Data/Zmod/Defs.lean]: > I can take a look at fixing them instead, though probably after CommRing works

I don't understand this comment; the code that I'm suggesting you throw away is the CommRing code you wrote / took from the ad-hoc port. We should put the mathport CommRing code back and then try and fix the errors it creates.

Comment (qawbecrdtey) [reply Mathlib/Data/Zmod/Defs.lean]: Removing curly braces at line 50 and 59 worked when I was working on 
[commit fd60ee036a8a9b33183779cce999dec6f24f43d5](https://github.com/leanprover-community/mathlib4/pull/1684/commits/fd60ee036a8a9b33183779cce999dec6f24f43d5).

Comment (Ruben-VandeVelde) [reply Mathlib/Data/Zmod/Defs.lean]: Oh, I'm sorry for the confusion. It's `CommRing (Zmod n)` that's the main blocker here, not anything about `Fin`

Comment (eric-wieser) [reply Mathlib/Data/Zmod/Defs.lean]: I've pushed a fix to that instance

**Comment by eric-wieser** [comment]
Was the Zulip consensus to use `Zmod` or `ZMod`?

File: Mathlib/Data/Zmod/AdHocDefs.lean [hunk 3]
```diff
@@ -1,72 +0,0 @@
...
-section CommRing
-/-- Modular projection `ℤ → Fin n` whenever `Fin n` is nonempty. -/
-protected def ofInt'' [NeZero n] : Int → Fin n
-  | Int.ofNat a => Fin.ofNat' a Fin.size_positive'
-  | Int.negSucc a => -(Fin.ofNat' a.succ Fin.size_positive')
-
-/-- Modular projection `ℤ → Fin n` whenever `Fin n` is nonempty. -/
-def ofInt' [NeZero n] : ℤ → Fin n
-  | (i : ℕ) => i
-  | (Int.negSucc i) => -↑(i + 1 : ℕ)
```
Comment (eric-wieser) [inline Mathlib/Data/Zmod/AdHocDefs.lean:56]: This was an adhoc refactor that never made it to Mathlib3. This probably is a better implementation of `ofInt`, but only very marginally. If someone felt strongly about this being useful, they would have backported it to mathlib3 a long time ago.

We can always revisit after the port.

File: Mathlib/Data/Zmod/Defs.lean [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+instance Zmod.decidableEq : ∀ n : ℕ, DecidableEq (Zmod n)
+  | 0 => by dsimp [Zmod]; infer_instance
+  | n + 1 => by dsimp [Zmod]; infer_instance
+#align zmod.decidable_eq Zmod.decidableEq
+
+instance Zmod.hasRepr : ∀ n : ℕ, Repr (Zmod n)
+  | 0 => by dsimp [Zmod]; infer_instance
+  | n + 1 => by dsimp [Zmod]; infer_instance
+#align zmod.has_repr Zmod.hasRepr
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Zmod/Defs.lean:100]:
```suggestion
instance Zmod.repr : ∀ n : ℕ, Repr (Zmod n)
  | 0 => by dsimp [Zmod]; infer_instance
  | n + 1 => by dsimp [Zmod]; infer_instance
#align zmod.has_repr Zmod.repr
```

**Comment by Ruben-VandeVelde** [comment]
Thanks!

File: Mathlib/Data/Zmod/Defs.lean [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+This used to be inlined into data/zmod/basic.lean. This file imports `char_p/basic`, which is an
+issue; all `char_p` instances create an `algebra (zmod p) R` instance; however, this instance may
+not be definitionally equal to other `algebra` instances (for example, `galois_field` also has an
+`algebra` instance as it is defined as a `splitting_field`). The way to fix this is to use the
+forgetful inheritance pattern, and make `char_p` carry the data of what the `smul` should be (so
+for example, the `smul` on the `galois_field` `char_p` instance should be equal to the `smul` from
+its `splitting_field` structure); there is only one possible `zmod p` algebra for any `p`, so this
+is not an issue mathematically. For this to be possible, however, we need `char_p/basic` to be
+able to import some part of `zmod`.
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Zmod/Defs.lean:30]:
```suggestion
/-!
# Definition of `Zmod n` + basic results.

This file provides the basic details of `Zmod n`, including its commutative ring structure.

## Implementation details

This used to be inlined into `Data.Zmod.Basic`. This file imports `CharP.Basic`, which is an
issue; all `CharP` instances create an `Algebra (Zmod p) R` instance; however, this instance may
not be definitionally equal to other `Algebra` instances (for example, `GaloisField` also has an
`Algebra` instance as it is defined as a `SplittingField`). The way to fix this is to use the
forgetful inheritance pattern, and make `CharP` carry the data of what the `smul` should be (so
for example, the `smul` on the `GaloisField` `CharP` instance should be equal to the `smul` from
its `SplittingField` structure); there is only one possible `Zmod p` algebra for any `p`, so this
is not an issue mathematically. For this to be possible, however, we need `CharP.Basic` to be
able to import some part of `Zmod`.
```

File: Mathlib/Data/Zmod/Defs.lean [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+
+namespace Fin
+
+/-!
+## Ring structure on `fin n`
+
+We define a commutative ring structure on `fin n`, but we do not register it as instance.
+Afterwords, when we define `zmod n` in terms of `fin n`, we use these definitions
+to register the ring structure on `zmod n` as type class instance.
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Zmod/Defs.lean:42]:
```suggestion
## Ring structure on `Fin n`

We define a commutative ring structure on `Fin n`, but we do not register it as instance.
Afterwords, when we define `Zmod n` in terms of `Fin n`, we use these definitions
to register the ring structure on `Zmod n` as type class instance.
```

Comment (Ruben-VandeVelde) [reply Mathlib/Data/Zmod/Defs.lean:42]: (though this seems to be false anyway)

File: Mathlib/Data/Zmod/Defs.lean [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+-/
+
+
+open Nat.ModEq Int
+
+/-- Multiplicative commutative semigroup structure on `fin n`. -/
+instance (n : ℕ) : CommSemigroup (Fin n) :=
+  {
+    inferInstanceAs (Mul (Fin n)) with
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Zmod/Defs.lean:51]:
```suggestion
  { inferInstanceAs (Mul (Fin n)) with
```

File: Mathlib/Data/Zmod/Defs.lean [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+private theorem left_distrib_aux (n : ℕ) : ∀ a b c : Fin n, a * (b + c) = a * b + a * c :=
+  fun ⟨a, ha⟩ ⟨b, hb⟩ ⟨c, hc⟩ =>
+  Fin.eq_of_veq
+    (calc
+      a * ((b + c) % n) ≡ a * (b + c) [MOD n] := (Nat.mod_modEq _ _).mul_left _
+      _ ≡ a * b + a * c [MOD n] := by rw [mul_add]
+      _ ≡ a * b % n + a * c % n [MOD n] := (Nat.mod_modEq _ _).symm.add (Nat.mod_modEq _ _).symm
+      )
+
+/-- Commutative ring structure on `fin n`. -/
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Zmod/Defs.lean:70]:
```suggestion
/-- Commutative ring structure on `Fin n`. -/
```

File: Mathlib/Data/Zmod/Defs.lean (hunk 6 above)
Comment (jcommelin) [reply Mathlib/Data/Zmod/Defs.lean]: Yeah, feel free to rewrite.

File: Mathlib/Data/Zmod/Defs.lean [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+
+namespace Fin
+
+/-!
+## Ring structure on `Fin n`
+
+We define a commutative ring structure on `Fin n`, but we do not register it as instance.
+Afterwords, when we define `Zmod n` in terms of `Fin n`, we use these definitions
+to register the ring structure on `Zmod n` as type class instance.
```
Comment (jcommelin) [inline Mathlib/Data/Zmod/Defs.lean:42]: Feel free to rewrite, and maybe leave a TODO that we should streamline this file, now that `Fin` has more instances.

File: scripts/start_port.sh [hunk 10]
```diff
@@ -82,4 +82,3 @@ git commit \
 
 echo "After pushing, you can open a PR at:"
 echo "https://github.com/leanprover-community/mathlib4/compare/$branch_name?expand=1&title=feat:+port+$mathlib4_mod_tail"
-
```
Comment (jcommelin) [inline scripts/start_port.sh:85]: Is this intentional?

**Review by jcommelin — COMMENTED** [review]
bors d+

File: Mathlib/Data/Zmod/Defs.lean [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+      rw [mul_comm, left_distrib_aux, mul_comm _ b, mul_comm],
+    -- porting note: new, see
+    -- https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/ring.20vs.20Ring/near/322876462
+    zero_mul := Fin.zero_mul
+    mul_zero := Fin.mul_zero }
+
+end Fin
+
+/-- The integers modulo `n : ℕ`. -/
+def Zmod : ℕ → Type
```
Comment (eric-wieser) [inline Mathlib/Data/Zmod/Defs.lean:86]:
```suggestion
def ZMod : ℕ → Type
```
and throughout (including the filename), as requested by @hrmacbeth [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Naming.20convention/near/322551595)

File: Mathlib/Data/Zmod/Defs.lean (hunk 9 above)
Comment (Ruben-VandeVelde) [reply Mathlib/Data/Zmod/Defs.lean]: I just dropped the obviously false bit

**Review by eric-wieser — APPROVED** [review]
bors merge

Thanks!

**Pull request closed without merging**